cmd/zfs/zfs_main.c

```diff
@@ -727,6 +727,38 @@ finish_progress(char *done)
 	pt_header = NULL;
 }
 
+/* This function checks if the passed fd refers to /dev/null or /dev/zero */
+#ifdef __linux__
+static int
+is_dev_files(int fd)
+{
+	int nfd = open("/dev/null", O_WRONLY);
+	struct stat st, nst;
+	fstat(fd, &st);
+	fstat(nfd, &nst);
+	int res = (st.st_dev == nst.st_dev);
+	close(nfd);
+	if (res != 1) {
+		nfd = open("/dev/zero", O_WRONLY);
+		fstat(nfd, &nst);
+		res = (st.st_dev == nst.st_dev);
+		close(nfd);
+	}
+	return (res);
+}
+#endif
+
+static int
+is_dev_error(int err, int fd)
+{
+#ifdef __linux__
+	if (err == EINVAL && is_dev_files(fd)) {
```

Comment:
`return err == EINVAL && is_dev_nullzero(fd);`

Reply:
Is your suggestion that I drop the `#ifdef __linux__`, then?

Reply:
Not necessarily; something like
```c
#ifdef __linux__
return (err == EINVAL && is_dev_nullzero(fd));
#else
return (B_FALSE);
#endif
```
might be better here

Reply:
Sure, I just wanted to ask whether that was what you had in mind before pushing a change.

How's that?

Reply:
looks great to me